### ISSUE TITLE
render novel_pose when using train_stage=2

### ISSUE THREAD

**Issue by Xzy765039540** [huliangxiao/GaussianAvatar#8]
How to render novel_pose when using train_stage=2? Can you provide a script? 
Best Regards,

**Comment by wyiguanw** [comment]
hi, i met same issue. and I try to modify `render_novel_pose.py` for stage 2, here is the result:
![00000](https://github.com/huliangxiao/GaussianAvatar/assets/95457754/254360c2-5805-44f3-8d13-52d64d5af34c)

but the visualization during the training is fine:
 
![08001_pred](https://github.com/huliangxiao/GaussianAvatar/assets/95457754/3729a6c4-e484-438d-b045-e8801f5ab8f8)


this script works perfectly for stage 1, but not for stage 2. and the only different thing that I found is stage 2 have a pose_featmap so, Im not sure why this happened. could you provide script for stage 2 rendering

**Comment by Xzy765039540** [comment]
> hi, i met same issue. and I try to modify `render_novel_pose.py` for stage 2, here is the result: ![00000](https://private-user-images.githubusercontent.com/95457754/304585305-254360c2-5805-44f3-8d13-52d64d5af34c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgyMzgxNTMsIm5iZiI6MTcwODIzNzg1MywicGF0aCI6Ii85NTQ1Nzc1NC8zMDQ1ODUzMDUtMjU0MzYwYzItNTgwNS00NGYzLThkMTMtNTJkNjRkNWFmMzRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE4VDA2MzA1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1ODI0ODRkN2IwMDM2ZThhMmQ1ZTA1ODUzMmIyN2UxZmNiNGIxMGNhYjQxNzVjNDZmN2U2ZDIxOWE0NzdiZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mO1ptSIKRyiK3TYmJFoTQz-VRzRXs5aoXaOO0ZmD4RI)
> 
> but the visualization during the training is fine:
> 
> ![08001_pred](https://private-user-images.githubusercontent.com/95457754/304585444-3729a6c4-e484-438d-b045-e8801f5ab8f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgyMzgxNTMsIm5iZiI6MTcwODIzNzg1MywicGF0aCI6Ii85NTQ1Nzc1NC8zMDQ1ODU0NDQtMzcyOWE2YzQtZTQ4NC00MzhkLWIwNDUtZTg4MDFmNWFiOGY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE4VDA2MzA1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2NGVjZmU4MDM0MWY0OTlmMjkwNWZjMTcyNzdkN2I4MjAyOGYxZjEyYTJlZDBhOGZjZGNmMmFmM2M0MTI4MzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.13wbbvD8PgpGdkU0JoH2LcYAaZrhNeCRa4v3CJ_p7DE)
> 
> this script works perfectly for stage 1, but not for stage 2. and the only different thing that I found is stage 2 have a pose_featmap so, Im not sure why this happened. could you provide script for stage 2 rendering

I think the main reason should be that the pose_featmap is not loaded when the model is loaded, I slove this problem by add self.pose_encoder.load_state_dict(saved_model_state["pose_encoder"], strict=False) when use stage 2 , I modified the load() function in avatar_model.py
def load(self, iteration, test=False):

        net_save_path = os.path.join(self.model_path, "net/iteration_{}".format(iteration))
        if self.model_parms.train_stage  ==1:
            saved_model_state = torch.load(
                os.path.join(net_save_path, "net.pth"))
            print('load pth: ', os.path.join(net_save_path, "net.pth"))
            self.net.load_state_dict(saved_model_state["net"], strict=False)
        
        if self.model_parms.train_stage  ==2:
            saved_model_state = torch.load(
                os.path.join(net_save_path, "pose_encoder.pth"))
            print('load pth: ', os.path.join(net_save_path, "pose_encoder.pth"))
            self.net.load_state_dict(saved_model_state["net"], strict=False)

        if self.model_parms.train_stage  ==1:
            if not test:
                self.pose.load_state_dict(saved_model_state["pose"], strict=False)
                self.transl.load_state_dict(saved_model_state["transl"], strict=False)
            # if self.train_mode == 0:
            self.geo_feature.data[...] = saved_model_state["geo_feature"].data[...]
            
        if self.model_parms.train_stage  ==2:
            if not test:
                self.pose.load_state_dict(saved_model_state["pose"], strict=False)
                self.transl.load_state_dict(saved_model_state["transl"], strict=False)
            # if self.train_mode == 0:
            self.geo_feature.data[...] = saved_model_state["geo_feature"].data[...]
            self.pose_encoder.load_state_dict(saved_model_state["pose_encoder"], strict=False)

        if self.optimizer is not None:
            self.optimizer.load_state_dict(saved_model_state["optimizer"])
        if self.scheduler is not None:
            self.scheduler.load_state_dict(saved_model_state["scheduler"])

**Comment by Xzy765039540** [comment]
May I ask what the main improvements were made in stage 2, because I've noticed that there's not much difference in the rendering results between stage 1 and stage 2.

**Comment by serizawa-04013958** [comment]
I faced same problems. How can i render with stage2?
my stage1 results with neuman dataset was not so good.
![00212](https://github.com/aipixel/GaussianAvatar/assets/78073281/b2443feb-f1c2-4fed-bef4-404b29964def)

**Comment by silkA111** [comment]
> > 嗨，我遇到了同样的问题。我尝试修改第 2 阶段，结果如下：`render_novel_pose.py`![00000](https://private-user-images.githubusercontent.com/95457754/304585305-254360c2-5805-44f3-8d13-52d64d5af34c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgyMzgxNTMsIm5iZiI6MTcwODIzNzg1MywicGF0aCI6Ii85NTQ1Nzc1NC8zMDQ1ODUzMDUtMjU0MzYwYzItNTgwNS00NGYzLThkMTMtNTJkNjRkNWFmMzRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE4VDA2MzA1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1ODI0ODRkN2IwMDM2ZThhMmQ1ZTA1ODUzMmIyN2UxZmNiNGIxMGNhYjQxNzVjNDZmN2U2ZDIxOWE0NzdiZmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mO1ptSIKRyiK3TYmJFoTQz-VRzRXs5aoXaOO0ZmD4RI)
> > 但是训练期间的可视化效果很好：
> > ![08001_pred](https://private-user-images.githubusercontent.com/95457754/304585444-3729a6c4-e484-438d-b045-e8801f5ab8f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgyMzgxNTMsIm5iZiI6MTcwODIzNzg1MywicGF0aCI6Ii85NTQ1Nzc1NC8zMDQ1ODU0NDQtMzcyOWE2YzQtZTQ4NC00MzhkLWIwNDUtZTg4MDFmNWFiOGY4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE4VDA2MzA1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2NGVjZmU4MDM0MWY0OTlmMjkwNWZjMTcyNzdkN2I4MjAyOGYxZjEyYTJlZDBhOGZjZGNmMmFmM2M0MTI4MzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.13wbbvD8PgpGdkU0JoH2LcYAaZrhNeCRa4v3CJ_p7DE)
> > 此脚本适用于第 1 阶段，但不适用于第 2 阶段。我发现的唯一不同之处在于第 2 阶段有一个pose_featmap所以，我不确定为什么会这样。您能否提供第 2 阶段渲染的脚本
> 
> 我认为主要原因应该是加载模型时没有加载pose_featmap，我在使用阶段 2 时通过添加 self.pose_encoder.load_state_dict（saved_model_state[“pose_encoder”]， strict=False） 来解决这个问题，我修改了 avatar_model.py def load（self， iteration， test=False） 中的 load（） 函数：
> 
> ```
>     net_save_path = os.path.join(self.model_path, "net/iteration_{}".format(iteration))
>     if self.model_parms.train_stage  ==1:
>         saved_model_state = torch.load(
>             os.path.join(net_save_path, "net.pth"))
>         print('load pth: ', os.path.join(net_save_path, "net.pth"))
>         self.net.load_state_dict(saved_model_state["net"], strict=False)
>     
>     if self.model_parms.train_stage  ==2:
>         saved_model_state = torch.load(
>             os.path.join(net_save_path, "pose_encoder.pth"))
>         print('load pth: ', os.path.join(net_save_path, "pose_encoder.pth"))
>         self.net.load_state_dict(saved_model_state["net"], strict=False)
> 
>     if self.model_parms.train_stage  ==1:
>         if not test:
>             self.pose.load_state_dict(saved_model_state["pose"], strict=False)
>             self.transl.load_state_dict(saved_model_state["transl"], strict=False)
>         # if self.train_mode == 0:
>         self.geo_feature.data[...] = saved_model_state["geo_feature"].data[...]
>         
>     if self.model_parms.train_stage  ==2:
>         if not test:
>             self.pose.load_state_dict(saved_model_state["pose"], strict=False)
>             self.transl.load_state_dict(saved_model_state["transl"], strict=False)
>         # if self.train_mode == 0:
>         self.geo_feature.data[...] = saved_model_state["geo_feature"].data[...]
>         self.pose_encoder.load_state_dict(saved_model_state["pose_encoder"], strict=False)
> 
>     if self.optimizer is not None:
>         self.optimizer.load_state_dict(saved_model_state["optimizer"])
>     if self.scheduler is not None:
>         self.scheduler.load_state_dict(saved_model_state["scheduler"])
> ```
Have you solved it? How to render new postures in the second stage after training
